src/main.rs

```diff
@@ -1,26 +1,51 @@
 #![allow(clippy::module_inception)]
 
+mod error;
 mod ice;
 mod lexer;
 mod parser;
+mod source_file;
 
+use crate::error::{print_error, CompilerError};
+use crate::parser::ast::Ast;
+use crate::source_file::SourceFile;
 use bumpalo::Bump;
 use clap::{self, Parser as ClapParser};
+use lexer::token_buffer::TokenBuffer;
 use parser::Parser;
+use std::process::ExitCode;
 
 #[derive(ClapParser)]
 #[command(author, version, about, long_about = None)]
 struct Options {
     file_name: String,
 }
 
-fn main() -> miette::Result<()> {
+fn compile<'s, 'a>(source: SourceFile<'s>, arena: &'a Bump) -> Result<Ast<'s, 'a>, CompilerError> {
+    let token_buffer = TokenBuffer::from_source(source)?;
+
+    let parser = Parser::new(token_buffer, arena);
+    let ast = parser.parse()?;
+
+    Ok(ast)
+}
+
+fn main() -> ExitCode {
```

Comment:
Why not a `Result` and let that handle the exit code with `Termination` automatically?